### PR TITLE
Fix KubernetesHook failed on an attribute metadata:name absence

### LINKED ISSUE #25668: SparkKubernetesOperator application file attribute "name" is not mandatory
### Apache Airflow version

2.3.3

### What happened

Since commit https://github.com/apache/airflow/commit/3c5bc73579080248b0583d74152f57548aef53a2 the SparkKubernetesOperator application file is expected to have an attribute metadata:name and the operator execution fails with exception `KeyError: 'name'` if it not exists. Please find the example error stack below:

```
[2022-07-27, 12:58:07 UTC] {taskinstance.py:1909} ERROR - Task failed with exception
Traceback (most recent call last):
  File "/opt/bitnami/airflow/venv/lib/python3.8/site-packages/airflow/providers/cncf/kubernetes/operators/spark_kubernetes.py", line 69, in execute
    response = hook.create_custom_object(
  File "/opt/bitnami/airflow/venv/lib/python3.8/site-packages/airflow/providers/cncf/kubernetes/hooks/kubernetes.py", line 316, in create_custom_object
    name=body_dict["metadata"]["name"],
KeyError: 'name'
```

### What you think should happen instead

The operator should start successfully, ignoring the field absence
The attribute metadata:name in NOT mandatory, and a pair metadata:generateName can be user instead - please find proof here: https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.22/#objectmeta-v1-meta, particularly in the following:

```
GenerateName is an optional prefix, used by the server, to generate a unique name ONLY IF the Name field has not been provided
```

### How to reproduce

Start a DAG with SparkKubernetesOperator with an application file like this in the beginning:

```
apiVersion: sparkoperator.k8s.io/v1beta2
kind: SparkApplication
metadata:
  generateName: spark_app_name
[...]
```

### Operating System

linux

### Versions of Apache Airflow Providers

apache-airflow==2.3.3
apache-airflow-providers-cncf-kubernetes==4.2.0

### Deployment

Virtualenv installation

### Deployment details

_No response_

### Anything else

Every time

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by morooshka**
closes: #25668
related: #25668

### CODE REVIEW

**Comment by potiuk** [comment]
There are errors to fix.

**Comment by morooshka** [comment]
> There are errors to fix.
@potiuk thank you for your message. I see some tests failing, but I consider these are not connected with my changes. How can I resolve the case? I have made several rebases on apache/main and the number of falling tests changes from iteration to iteration. What is the best practice with this? Thank you

**Comment by morooshka** [comment]
@potiuk The problem had gone. I was so mind fixed that it is not connected with my changes that I really missed a piece of code deleted by occasion. Thank you, you forced me to recheck

**Comment by potiuk** [comment]
> @potiuk The problem had gone. I was so mind fixed that it is not connected with my changes that I really missed a piece of code deleted by occasion. Thank you, you forced me to recheck

:D -> yeah complexity of our CI leads it to some false negatives that might cloud the understanding which changes are causing which problems. Actually - we have a main failure currently that needs to be fixed that WILL likely fail in your PR so brace for it. 

I wish we had the situaton that we have super-stable and solid CI where we have 100% certaintty about the roor cause of the problem but this is IMHO very nice goal to have but very unrealistic to achieve in practice for any system which has sizeable complexity.  It's pretty much always a moving target and we need to accept it as a fact of life (and strive to improve it continuousy without the false hopes we will ever achieve it "for ever". There are brief periods where it might work and the most we can achieve is to make those periods longere and more "stable".

**Comment by morooshka** [comment]
@potiuk Sure, I understand the mission =)

**Comment by potiuk** [comment]
And now. -you have some conflicts to solve and rebase, look at the Helm isssues and see if those are your problems and fix them if nedeed.

**Comment by morooshka** [comment]
> And now. -you have some conflicts to solve and rebase, look at the Helm isssues and see if those are your problems and fix them if nedeed.

Hi Jarek! Fixed. Some CI pipes were skipped, I have reported to Slack. I do not understand why and do not understand how to get the reason. However, the checks were successful 😁

**Comment by potiuk** [comment]
Let's see. we skip some unnneded checks (based on what your change brings) but I also have to approve the rest of your workflow because you are "new' user.

**Comment by potiuk** [comment]
So some checks will still run

**Comment by morooshka** [comment]
> Let's see. we skip some unnneded checks (based on what your change brings) but I also have to approve the rest of your workflow because you are "new' user.

The skipped checks we reported as failed by email to me:
image

**Comment by potiuk** [comment]
> > Let's see. we skip some unnneded checks (based on what your change brings) but I also have to approve the rest of your workflow because you are "new' user.
> 
> The skipped checks we reported as failed by email to me: image

Those are likely tests that are run in your fork "main" after synchronizing the changes. You should likely disable the workflows in your own fork.

**Comment by potiuk** [comment]
All our tests for PRs from fork run in Airflow repo's GitHub Actions

**Comment by morooshka** [comment]
> So some checks will still run

Hi Jarek! Finally all checks succeeded. Do I need to rebase again? The main head had gone away again =)

**Comment by potiuk** [comment]
No. It's ok. But I think @jedcunningham aand @dstandish / @ephraimbuddy shoudl take a look at it :)

**Review by potiuk — APPROVED** [review]
LGTM. But I woudl love others to confirm :)

File: airflow/providers/cncf/kubernetes/hooks/kubernetes.py [hunk 1]
```diff
@@ -301,36 +301,39 @@ def create_custom_object(
...
-                name=body_dict["metadata"]["name"],
-            )
-            self.log.warning("Deleted SparkApplication with the same name.")
-        except client.rest.ApiException:
-            self.log.info("SparkApp %s not found.", body_dict['metadata']['name'])
+
+        body_dict: dict = _load_body_to_dict(body) if isinstance(body, str) else body
+
+        # Attribute "name" is not mandatory if "generateName" is used instead
+        if "name" in body_dict["metadata"].keys():
```
Comment (mik-laj) [inline airflow/providers/cncf/kubernetes/hooks/kubernetes.py]:
```suggestion
        if "name" in body_dict["metadata"]:
```

Comment (morooshka) [reply airflow/providers/cncf/kubernetes/hooks/kubernetes.py]: Hi Kamil! Thank you for joining reviewing my PR. I use exact .keys() mention because its more definitely.  Is it bad style?

Comment (dstandish) [reply airflow/providers/cncf/kubernetes/hooks/kubernetes.py]: yeah you don't need to, it's commonly accepted to do `if key in dict`

File: airflow/providers/cncf/kubernetes/hooks/kubernetes.py [hunk 2]
```diff
@@ -301,36 +301,39 @@ def create_custom_object(
...
-                name=body_dict["metadata"]["name"],
-            )
-            self.log.warning("Deleted SparkApplication with the same name.")
-        except client.rest.ApiException:
-            self.log.info("SparkApp %s not found.", body_dict['metadata']['name'])
+
+        body_dict: dict = _load_body_to_dict(body) if isinstance(body, str) else body
+
+        # Attribute "name" is not mandatory if "generateName" is used instead
+        if "name" in body_dict["metadata"]:
```
Comment (dstandish) [inline airflow/providers/cncf/kubernetes/hooks/kubernetes.py:269]: can you clarify what you're doing here with this change?

you say
> Attribute "name" is not mandatory if "generateName" is used instead

which suggests, you can still delete the object without `name` as long as `generateName` is present.

but, with your change, if `name` is not present, it appears you won't attempt to delete the object at all!

is it not required to delete the object here? am i missing something?

**Review by dstandish — APPROVED** [review]

File: airflow/providers/cncf/kubernetes/hooks/kubernetes.py (hunk 2 above)
Comment (morooshka) [reply airflow/providers/cncf/kubernetes/hooks/kubernetes.py:269]: Hi Daniel! Recently there was no functionality of removing already launched objects. So no deleting is ok. Then it was implemented, but the author thought the "name" attribute is mandatory. It is not true -  "generateName" can be used. So we have 2 cases:
1. "name" is used. OK, we can delete the old object if found
2. "name" is not used. It is possible if "generateName" is used. If so, we can not delete the old object - we do not know its name (k8s will generate it randomly using a name mask, but we cant guess it for sure). So we have nothing to do there, the old object will be not touched if exists
Hope this sounds understandable, please ask further to clear the case
Thank you

File: airflow/providers/cncf/kubernetes/hooks/kubernetes.py (hunk 1 above)
Comment (morooshka) [reply airflow/providers/cncf/kubernetes/hooks/kubernetes.py]: Daniel, thank you, I will follow that style. I see someone made a commit - so I do not need to make an additional patch?

**Comment by morooshka** [comment]
Hi @potiuk! How can we go further?

**Comment by potiuk** [comment]
Screenshot 2022-10-10 at 00 28 09

Start with resolving conflicts and ask here for merging. I am going for holidays now, so won't be able to merge it, but others might.

File: airflow/providers/cncf/kubernetes/hooks/kubernetes.py [hunk 3]
```diff
@@ -56,15 +56,15 @@ class KubernetesHook(BaseHook):
...
-    :param config_file: Path to kubeconfig file.
-    :param in_cluster: Set to ``True`` if running from within a kubernetes cluster.
-    :param disable_verify_ssl: Set to ``True`` if SSL verification should be disabled.
-    :param disable_tcp_keepalive: Set to ``True`` if you want to disable keepalive logic.
+        in your kubeconfig
+    :param config_file: Path to kubeconfig file
+    :param in_cluster: Set to ``True`` if running from within a kubernetes cluster
+    :param disable_verify_ssl: Set to ``True`` if SSL verification should be disabled
+    :param disable_tcp_keepalive: Set to ``True`` if you want to disable keepalive logic
     """
```
Comment (dstandish) [inline airflow/providers/cncf/kubernetes/hooks/kubernetes.py]: I appretiate the punctiliousness, however would you please revert these lines?  
Period at end of line in param docstring is not an established convention for airflow.  So it's not really worth the obfuscation of the git history.  Adding these kinds of unrelated changes to PRs is a distraction for reviewers.  I don't mind it when it's something uncontroversial / obviously positive, like fixing a typo.  But here a developer could go either way and no one would care, so we shouldn't change it for no reason.

Same for the capitalization you've introduced in comments below.

File: airflow/providers/cncf/kubernetes/hooks/kubernetes.py [hunk 4]
```diff
@@ -126,8 +126,8 @@ def __init__(
 
         self._is_in_cluster: bool | None = None
 
-        # these params used for transition in KPO to K8s hook
-        # for a deprecation period we will continue to consider k8s settings from airflow.cfg
```
Comment (dstandish) [inline airflow/providers/cncf/kubernetes/hooks/kubernetes.py]: best to leave it alone

File: airflow/providers/cncf/kubernetes/hooks/kubernetes.py [hunk 5]
```diff
@@ -301,36 +301,39 @@ def create_custom_object(
...
         :param namespace: kubernetes namespace
         """
         api = client.CustomObjectsApi(self.api_client)
+
         if namespace is None:
             namespace = self.get_namespace()
-        if isinstance(body, str):
-            body_dict = _load_body_to_dict(body)
-        else:
-            body_dict = body
```
Comment (dstandish) [inline airflow/providers/cncf/kubernetes/hooks/kubernetes.py]: again here is a completely unrelated change that is purely stylistic.  in this situation i think it's best to just leave it alone.  oneline vs if / else is the stylistic choice of the original developer, and it's perfectly acceptable. meanwhile, every change has risk of introducing bug and it's also a distraction for the reviewer.

**Comment by dstandish** [comment]
i hesitated suggesting reverting your unrelated changes cus i didn't want to be too nitpicky but after looking again, i do think it's best to revert those changes if you wouldn't mind since you're going to be rebasing anyway.

File: airflow/providers/cncf/kubernetes/hooks/kubernetes.py (hunk 5 above)
Comment (morooshka) [reply airflow/providers/cncf/kubernetes/hooks/kubernetes.py]: I would like to object here. In such ternary operations the special construction is a good style. Also we need type hint, which in its original form will be not good:
```
         if isinstance(body, str):
            body_dict: dict = _load_body_to_dict(body)
        else:
            body_dict: dict = body
```

File: airflow/providers/cncf/kubernetes/hooks/kubernetes.py (hunk 3 above)
Comment (morooshka) [reply airflow/providers/cncf/kubernetes/hooks/kubernetes.py]: I have changed these due to pycharm weak warnings. And the pros of these just not to stop an eye on these week warnings

Comment (dstandish) [reply airflow/providers/cncf/kubernetes/hooks/kubernetes.py]: well it's certainly not worth debating so, as you wish.  but going forward you might consider disabling those specific warnings since... they will pop up all over the codebase

File: airflow/providers/cncf/kubernetes/hooks/kubernetes.py (hunk 5 above)
Comment (dstandish) [reply airflow/providers/cncf/kubernetes/hooks/kubernetes.py]: same here, as you wish

**Comment by dstandish** [comment]
just need to resolve conflicts here

File: newsfragments/21092.bugfix.rst [hunk 6]
```diff
@@ -0,0 +1 @@
+``KubernetesHook`` failed on a manifest field ``metadata:name`` absence.
```
Comment (eladkal) [inline newsfragments/21092.bugfix.rst]: providers releases don't use news fragment so we should remove this rst file.
The change log will be automatically created from the commit message

**Review by eladkal — DISMISSED** [review]

File: airflow/providers/cncf/kubernetes/hooks/kubernetes.py (hunk 3 above)
Comment (potiuk) [reply airflow/providers/cncf/kubernetes/hooks/kubernetes.py]: Yep. We have the rule, that if we want to follow certain standards, we enforce it with pre-commit so that it can be kept coinsistently across the codebase- if there is no rule for that, modifying code because "someone's" warnings are showing up is not a consistent approach and suffers the "works-for-me" syndrome. 

We should either turn such things in a rule (which I think we should not in this case) or refrain from changing the code to satisfy the "individual" rule. PyCharm has a way to disable such rules individually - you can store it in your project after checking out. Since we do not "enforce" common configuration for PyCharm/Intellij etc. we do not want to share IDE configuration - instead we opted for .gitignoring known tool configurations so whatever you change in your Pycharm's project config will not be committed to the repo - feel free to change it.

File: airflow/providers/cncf/kubernetes/hooks/kubernetes.py (hunk 5 above)
Comment (potiuk) [reply airflow/providers/cncf/kubernetes/hooks/kubernetes.py]: The rule of thumb here is that if possible - such changes should be separted out. This makes it easier for cherry-picking (in case we decide to cherry-pick your change to 2.4 branch it might cause unnecessary conflicts that we will have to resolve. So I suggest you split that change out into a separate PR.

This is basically the rule that everyone shoudl follow  - when you make a bugfix or change, avoid doing unrelated refactoring and separate it out to a separate PR. We are continiuously doing it and while it des not always work, it is something we strive for. 

This is much more important for Airlfow's and Core provider's code - because it allows cherry-picking and produces much better changelogs (we use commit messages to automatically generate changelog. So this is rather justified requirement that has very good reasons and I kindly ask  you to follow that.

File: newsfragments/21092.bugfix.rst (hunk 6 above)
Comment (potiuk) [reply newsfragments/21092.bugfix.rst]: Yep. No need for that (and it a bit harmful if it stays here as it will impact airflow's changelog)..

File: airflow/providers/cncf/kubernetes/hooks/kubernetes.py (hunk 2 above)
Comment (potiuk) [reply airflow/providers/cncf/kubernetes/hooks/kubernetes.py:269]: I tihnk it would be great if you add some of that context to commit message. While it links to the issue, it's always good to summarse the context in the commit message so that someone looking at the message will be able to understand the context (rather than having to read the whole issue).

**Comment by potiuk** [comment]
This one needs rebase/conflict resolution after applying string normalization (See slack and devlist for information). Ths might be good opportunity to remove the unrelated changes and move them to separate PR if needed.

File: airflow/providers/cncf/kubernetes/hooks/kubernetes.py (hunk 3 above)
Comment (morooshka) [reply airflow/providers/cncf/kubernetes/hooks/kubernetes.py]: OK, got it

File: airflow/providers/cncf/kubernetes/hooks/kubernetes.py (hunk 5 above)
Comment (morooshka) [reply airflow/providers/cncf/kubernetes/hooks/kubernetes.py]: OK, thank you for comments
I see how high communications skills required from you to rule this project =)
I do appreciate this

File: newsfragments/21092.bugfix.rst (hunk 6 above)
Comment (morooshka) [reply newsfragments/21092.bugfix.rst]: OK. But this was not clear from docs

**Comment by morooshka** [comment]
I followed all the recommendations above and made corrections

**Comment by potiuk** [comment]
I thinkn you need to solve conflicts now :( @eladkal Are you ok ?

File: newsfragments/21092.bugfix.rst (hunk 6 above)
Comment (eladkal) [reply newsfragments/21092.bugfix.rst]: It was mentioned in the issue template but you deleted its content and replaced it with the text you wrote in the issue :)

Comment (morooshka) [reply newsfragments/21092.bugfix.rst]: Sorry, I have clicked "request review" and something changed itself. I have deleted the news fragment from PR as you asked. Is it ok?

**Review by eladkal — COMMENTED** [review]
need to resolve conflicts

**Comment by morooshka** [comment]
@potiuk @eladkal conflicts resolved, it is being builded. In the previous try I have done the same but the main branch had succeeded to go further

**Comment by morooshka** [comment]
@potiuk Can you help with restarting skipped build stage please?

**Review by eladkal — APPROVED** [review]

**Review by potiuk — APPROVED** [review]
Very cool. Love the commit message
